src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+			"INSERT INTO blocks (pos, data) VALUES ($1, $2)", 2,
+			(const Oid *)writeFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "update_block",
+			"UPDATE blocks SET data = $1 WHERE pos = $2", 2,
+			(const Oid *)updateFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_block",
+			"SELECT data FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "delete_block",
+			"DELETE FROM blocks WHERE pos = $1::bigint", 1,
+			(const Oid *)readFormats);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	m_res = PQprepare(m_conn, "read_pos", "SELECT pos FROM blocks", 0, NULL);
+
+	if (PQresultStatus(m_res) != PGRES_COMMAND_OK)
+	       	throw BaseException(PQresultErrorMessage(m_res));
+
+	PQclear(m_res);
+
+	infostream << "Postgres: SQL statements prepared." << std::endl;
+}
+
+void Database_PostgreSQL::dbconnect(void)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	m_conn = PQconnectdb(m_conninfo.c_str());
+
+	if (PQstatus(m_conn) != CONNECTION_OK) 
+	       	throw BaseException(PQerrorMessage(m_conn));
+
+	infostream << "Postgres: Database connection made." << std::endl;
+}
+
+void Database_PostgreSQL::connectString(void)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	if (m_conninfo.length() > 1)
+		return;
+
+	try {
+		// username
+		if (m_conf.exists("pg_username")) {
+		    m_user = m_conf.get("pg_username");
+		} else {
+		    m_user = "minetest";
+		    infostream << "Postgres: Using default user, minetest.\n";
+		}
+
+		// Postgres host
+		if (m_conf.exists("pg_host"))
+		    m_host = m_conf.get("host");
+
+		// Port number
+		if (m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+		    m_port = m_conf.get("pg_port");
+		} else if (m_conf.exists("pg_port") && !m_conf.exists("pg_host")) {
+		    m_port = m_conf.get("pg_port");
+		    m_host = std::string("localhost");
+		} else if (!m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+		    m_port = "5432";
+		} else {
+		    infostream << "Postgres: Using socket connection.\n";
+		}
+
+		// password
+		if (m_conf.exists("pg_password"))
+		    m_password = m_conf.get("pg_password");
+		else
+		    infostream << "Using no password.\n";
+
+		// database name
+		if (m_conf.exists("pg_dbname")) {
+		    m_dbname = m_conf.get("pg_dbname");
+		} else {
+		    m_dbname = std::string("minetest");
+		    infostream << "Postges: Using default database name, minetest.\n";
+		}
+
+	} catch(SettingNotFoundException e) {
+		throw SettingNotFoundException("Set database in world.mt to use the pgsql backend");
+	}
+
+	m_conninfo = "user=" + m_user + " dbname=" + m_dbname;
```

Comment:
This test would seem to indicate that std::ostringstream is slow compared to other methods. Since this string is only created at start up it really wont count against game performance, but you might be interested in this link.

http://stackoverflow.com/questions/18892281/most-optimized-way-of-concatenation-in-strings